File: next/dweb-browser/MicroService/src/Sys/Dns/NativeFetch.cs

using DwebBrowser.MicroService.Http;

namespace DwebBrowser.MicroService.Core;

using FetchAdapter = Func<MicroModule, PureRequest, Task<PureResponse?>>;

public class NativeFetch
{
    public NativeFetch()
    {
    }

    public static AdapterManager<FetchAdapter> NativeFetchAdaptersManager = new();
}

public abstract partial class MicroModule
{
    static HttpClient httpClient = new HttpClient().Also(it =>
    {
        /// 默认禁用缓存
        it.DefaultRequestHeaders.CacheControl = new() { NoCache = true };
    });
    public async Task<PureResponse> NativeFetchAsync(PureRequest pureRequest)
    {
        foreach (var fetchAdapter in NativeFetch.NativeFetchAdaptersManager.Adapters)
        {
            var response = await fetchAdapter(this, pureRequest);

            if (response is not null)
            {
                return response;
            }
        }

        var httpResponse = await httpClient.SendAsync(pureRequest.ToHttpRequestMessage());
        return await httpResponse.ToPureResponseAsync();
    }
    public Task<PureResponse> NativeFetchAsync(URL url) =>
        NativeFetchAsync(new PureRequest(url.Href, IpcMethod.Get));

    public Task<PureResponse> NativeFetchAsync(Uri url) =>
        NativeFetchAsync(new PureRequest(url.ToString(), IpcMethod.Get));

    public Task<PureResponse> NativeFetchAsync(string url) =>
        NativeFetchAsync(new PureRequest(url, IpcMethod.Get));
}

File: next/kmp/core/src/commonMain/kotlin/org/dweb_browser/core/std/dns/NativeFetch.kt
package org.dweb_browser.core.std.dns

import io.ktor.http.Url
import org.dweb_browser.core.help.AdapterManager
import org.dweb_browser.core.module.MicroModule
import org.dweb_browser.helper.Debugger
import org.dweb_browser.pure.http.PureClientRequest
import org.dweb_browser.pure.http.PureMethod
import org.dweb_browser.pure.http.PureResponse
import org.dweb_browser.pure.http.defaultHttpPureClient

typealias FetchAdapter = suspend (remote: MicroModule.Runtime, request: PureClientRequest) -> PureResponse?

val debugFetch = Debugger("fetch")

val debugFetchFile = Debugger("fetch-file")

/**
 * file:/// => /usr & /sys as const
 * file://file.std.dweb/ => /home & /tmp & /share as userData
 */
val nativeFetchAdaptersManager = NativeFetchAdaptersManager()

class NativeFetchAdaptersManager : AdapterManager<FetchAdapter>() {

  private var client = defaultHttpPureClient

  class HttpFetch(private val manager: NativeFetchAdaptersManager) {
    val client get() = manager.client
    suspend operator fun invoke(request: PureClientRequest) = fetch(request)
    suspend operator fun invoke(url: String, method: PureMethod = PureMethod.GET) =
      fetch(PureClientRequest(method = method, href = url))

    val fetch = client::fetch
  }

  val httpFetch = HttpFetch(this)
}

val httpFetch = nativeFetchAdaptersManager.httpFetch

suspend fun MicroModule.Runtime.nativeFetch(request: PureClientRequest): PureResponse {
  for (fetchAdapter in nativeFetchAdaptersManager.adapters) {
    val response = fetchAdapter(this, request)
    if (response != null) {
      return response
    }
  }
  return nativeFetchAdaptersManager.httpFetch(request)
}

suspend inline fun MicroModule.Runtime.nativeFetch(url: Url) = nativeFetch(PureMethod.GET, url)

suspend inline fun MicroModule.Runtime.nativeFetch(url: String) = nativeFetch(PureMethod.GET, url)

suspend inline fun MicroModule.Runtime.nativeFetch(method: PureMethod, url: Url) =
  nativeFetch(PureClientRequest(url.toString(), method))

suspend inline fun MicroModule.Runtime.nativeFetch(method: PureMethod, url: String) =
  nativeFetch(PureClientRequest(url, method))
